lib/include/openturns/OTtestcode.hxx

```diff
@@ -454,6 +429,596 @@ void assert_equal(const T & a, const T & b, const String errMsg = "")
+      const Scalar pdf1 = distributionClone.computePDF(x);
+      Point param2(parameter);
+      param2[j] -= epsilon;
+      distributionClone.setParameter(param2);
+      const Scalar pdf2 = distributionClone.computePDF(x);
+      pdfgrfd[j] = (pdf1 - pdf2) / (2.0 * epsilon);
+      
+    }
+    LOGTRACE(OSS() << "pdfgrad(fd)=" << pdfgrfd.__str__());
+    assert_almost_equal(pdfGr, pdfgrfd, 1e-5, 1e-5, "wrong pdf gradient for " + distribution_.__repr__());
+  }
+  
+  void checkCDFGrad() const
+  {
+    if (!distribution_.isContinuous())
+      return;
+    if (distribution_.getDimension() < 4)
+    {
+      LOGTRACE(OSS() << "checking CDF gradient...");
+      const Point x(distribution_.getRealization());
+      const Point parameter(distribution_.getParameter());
+      const Point cdfGr = distribution_.computeCDFGradient(x);
+      LOGTRACE(OSS() << "cdfgrad=    " << cdfGr.__str__() << " x=" << x.__str__() << " params=" << parameter.__str__());
+      if (cdfGr.getDimension() != parameter.getDimension())
+        throw TestFailed(OSS() << "wrong cdfGradient(x) dimension (" << cdfGr.getDimension() << ") expected (" << parameter.getDimension() << ") for " << distribution_);
+      const Scalar epsilon = std::pow(ResourceMap::GetAsScalar("DistFunc-Precision"), 1.0 / 3.0);
+      Point cdfgrfd(parameter.getDimension());
+      for (UnsignedInteger j = 0; j < parameter.getDimension(); ++ j)
+      {
+        Distribution distributionClone(distribution_);
+        Point param1(parameter);
+        param1[j] += epsilon;
+        distributionClone.setParameter(param1);
+        const Scalar cdf1 = distributionClone.computeCDF(x);
+        Point param2(parameter);
+        param2[j] -= epsilon;
+        distributionClone.setParameter(param2);
+        const Scalar cdf2 = distributionClone.computeCDF(x);
+        cdfgrfd[j] = (cdf1 - cdf2) / (2.0 * epsilon);
+      }
+      LOGTRACE(OSS() << "cdfgrad(fd)=" << cdfgrfd.__str__());
+      assert_almost_equal(cdfGr, cdfgrfd, 1e-5, 1e-5, "wrong cdf gradient for " + distribution_.__repr__());
+    }
+  }
+
+  void checkSampling() const
+  {
+    const Point x = distribution_.getRealization();
+    LOGTRACE(OSS() << "oneRealization=" << x);
+    if (x.getDimension() != distribution_.getDimension())
+      throw TestFailed(OSS() << "wrong realization dimension for " << distribution_);
+    
+    if (distribution_.isContinuous())
+    {
+      if (!distribution_.getRange().contains(x))
+        throw TestFailed(OSS() << "realization not in range for " << distribution_);
+    }
+    else if (distribution_.isDiscrete())
+    {
+      const Sample support(distribution_.getSupport());
+      if (support.find(x) >= support.getSize())
+        throw TestFailed(OSS() << "realization not in support for " << distribution_);
+    }
+
+    if (distribution_.getDimension() == 1)
+    {
+      LOGTRACE(OSS() << "checking sampling with Kolmogorov/ChiSquared ...");
+      UnsignedInteger size = 100;
```

Comment:
This magic constant is a moderate sample size. Please use an attribute to hold it.